File: src/main/java/org/bk/graphql/domain/Book.java
package org.bk.graphql.domain;

import org.bk.graphql.domain.validation.BookName;
import org.bk.graphql.domain.validation.PageCount;
import org.immutables.value.Value;

import java.time.Instant;
import java.util.List;

@Value.Immutable
public interface Book {
    static Book create(
            BookId id,
            String name,
            List<AuthorId> authors,
            int pageCount,
            Instant createdAt,
            Instant updatedAt,
            int version
    ) {
        return ImmutableBook.builder()
                .id(id)
                .name(BookName.of(name).value())
                .authors(authors.stream().distinct().toList())
                .pageCount(PageCount.of(pageCount).value())
                .createdAt(createdAt)
                .updatedAt(updatedAt)
                .version(version)
                .build();
    }

    BookId id();
    String name();
    List<AuthorId> authors();
    int pageCount();
    Instant createdAt();
    Instant updatedAt();
    int version();
}

File: src/main/kotlin/org/bk/graphql/db/Book.kt
package org.bk.graphql.db

import org.springframework.data.annotation.Id
import org.springframework.data.annotation.Version
import org.springframework.data.relational.core.mapping.Column
import org.springframework.data.relational.core.mapping.Table

@Table("books")
data class Book(
    @Id val id: String,
    @Column("name")
    val name: String,
    @Column("page_count")
    val pageCount: Int,
    @Column("author_id")
    val authorId: String,
    @Version val version: Int = 0)
